NeoML/src/TraditionalML/KMeansClustering.cpp

```diff
@@ -55,9 +55,729 @@ static CPtr<CDnnBlob> createWeightBlob( IMathEngine& mathEngine, const IClusteri
+		( ( CKMeansClustering::IThreadTask* )ptr )->CallRun( threadIndex );
+	} );
+}
+
+static void vectorEltwiseMultiply( const CKMeansClustering& owner, IMathEngine& mathEngine, int vectorSize,
+	const CConstFloatHandle& first, const CConstFloatHandle& second, const CFloatHandle& result )
+{
+	if( !isThreadTaskRelevant( vectorSize ) ) {
+		mathEngine.VectorEltwiseMultiply( first, second, result, vectorSize );
+		return;
+	}
+
+	CKMeansClustering::CVectorEltwiseMultiplyThreadTask task( owner,
+		vectorSize, mathEngine, first, second, result );
+
+	NEOML_NUM_THREADS( *owner.GetThreadPool(), &task, []( int threadIndex, void* ptr ) {
+		( ( CKMeansClustering::IThreadTask* )ptr )->CallRun( threadIndex, /*align*/16 );
+	} );
+}
+
+static void vectorMultichannelLookupAndAddToTable( const CKMeansClustering& owner,
+	IMathEngine& mathEngine, int vectorSize, int channelCount, const CConstIntHandle& input,
+	const CFloatHandle* lookupHandles, const CLookupDimension* lookupDimensions, int lookupCount,
+	const CConstFloatHandle& mult, const CConstFloatHandle& matrix, int outputChannels )
+{
+	if( !isThreadTaskRelevant( vectorSize ) ) {
+		mathEngine.VectorMultichannelLookupAndAddToTable( vectorSize, channelCount,
+			 input, lookupHandles, lookupDimensions, lookupCount, mult, matrix, outputChannels );
+		return;
+	}
+
+	NeoAssert( channelCount == 1 );
+	CKMeansClustering::CVectorMultichannelLookupAndAddToTableThreadTask task( owner, vectorSize,
+		mathEngine, input, lookupHandles, lookupDimensions, lookupCount, mult, matrix, outputChannels );
+
+	NEOML_NUM_THREADS( *owner.GetThreadPool(), &task, []( int threadIndex, void* ptr ) {
+		( ( CKMeansClustering::IThreadTask* )ptr )->CallRun( threadIndex );
+	} );
+}
+
+static void matrixMultiplyByTransposed(
+	const CKMeansClustering& owner, IMathEngine& mathEngine, int batchSize,
+	const CConstFloatHandle& first, int firstHeight, int firstWidth,
+	const CConstFloatHandle& second, int secondHeight,
+	const CFloatHandle& result, int resultBufferSize )
+{
+	NeoAssert( batchSize == 1 );
+	NeoAssert( resultBufferSize >= firstHeight * secondHeight );
+
+	if( !isThreadTaskRelevant( firstWidth * firstHeight * secondHeight ) )
+	{
+		mathEngine.MultiplyMatrixByTransposedMatrix(
+			first, firstHeight, firstWidth, firstWidth/*firstRowSize*/,
+			second, secondHeight, firstWidth/*secondRowSize*/,
+			result, secondHeight/*resultRowSize*/, 0/*resultBufferSize*/ );
+		return;
+	}
+
+	struct CThreadTaskMxMT final {
+		CThreadTaskMxMT( int threadCount, IMathEngine& mathEngine,
+				const CConstFloatHandle& first, int firstHeight, int firstWidth,
+				const CConstFloatHandle& second, int secondHeight,
+				const CFloatHandle& result ) :
+			ThreadCount( threadCount ),
+			MathEngine( mathEngine ),
+			First( first ),
+			FirstHeight( firstHeight ),
+			FirstWidth( firstWidth ),
+			Second( second ),
+			SecondHeight( secondHeight ),
+			Result( result )
+		{}
+
+		void CallRun( int threadIndex )
+		{
+			int firstHeightStart;
```

Comment:
1. Multiline method inside the class - not good (by our codestyle)
2. 40-line `struct CThreadTaskMxMT` inside of method - also not good
Better keep it outside of a function. And `CThreadTaskMxMT::CallRun` outside of a class.

Reply:
fixed